### PR TITLE
add validator for x509 credential attributes

### LINKED ISSUE #3599: X509Credential validation
According to RFC, does credentialSubject contain only properties from the did:x509

### PR DESCRIPTION
**Pull request by woutslakhorst**
closes #3599

### CODE REVIEW

File: vcr/credential/validator.go [hunk 1]
```diff
@@ -253,3 +256,89 @@ func validateNutsCredentialID(credential vc.VerifiableCredential) error {
...
+	}
+	for _, policy := range policies[1:] {
+		policySplit := strings.Split(policy, ":")
+		if len(policySplit) < 2 {
+			return fmt.Errorf("invalid did:x509 policy '%s'", policy)
+		}
+		policyName := policySplit[0]
+		policyMap[policyName] = make(map[string]string)
+		for i := 1; i < len(policySplit); i += 2 {
+			unscaped, _ := url.PathUnescape(policySplit[i+1])
```
Comment (gerardsn) [inline vcr/credential/validator.go:312]: panics if `[i+1]` is out of bounds

File: vcr/credential/validator.go [hunk 2]
```diff
@@ -253,3 +256,89 @@ func validateNutsCredentialID(credential vc.VerifiableCredential) error {
...
+	}
+	x509resolver := didx509.NewResolver()
+	resolveMetadata := resolver.ResolveMetadata{}
+	if credential.Format() == vc.JWTCredentialProofFormat {
+		headers, err := crypto.ExtractProtectedHeaders(credential.Raw())
+		if err != nil {
+			return fmt.Errorf("%w: invalid JWT headers: %w", errValidation, err)
+		}
+		resolveMetadata.JwtProtectedHeaders = headers
+	}
```
Comment (gerardsn) [inline vcr/credential/validator.go:277]: add `else` with helpful error message

File: vcr/credential/validator.go [hunk 3]
```diff
@@ -253,3 +256,89 @@ func validateNutsCredentialID(credential vc.VerifiableCredential) error {
...
+		return fmt.Errorf("%w: %w", errValidation, err)
+	}
+
+	return (defaultCredentialValidator{}).Validate(credential)
+}
+
+// validatePolicyAssertions checks if the credentialSubject claims match the did issuer policies
+func validatePolicyAssertions(credential vc.VerifiableCredential) error {
+	// get base form of all credentialSubject
+	var target = make([]map[string]interface{}, 0)
```
Comment (gerardsn) [inline vcr/credential/validator.go:293]: micro optimize for expected length
```suggestion
	var target = make([]map[string]interface{}, 1)
```

File: vcr/credential/validator.go [hunk 4]
```diff
@@ -253,3 +256,89 @@ func validateNutsCredentialID(credential vc.VerifiableCredential) error {
...
+
+	if err = validatePolicyAssertions(credential); err != nil {
+		return fmt.Errorf("%w: %w", errValidation, err)
+	}
+
+	return (defaultCredentialValidator{}).Validate(credential)
+}
+
+// validatePolicyAssertions checks if the credentialSubject claims match the did issuer policies
+func validatePolicyAssertions(credential vc.VerifiableCredential) error {
```
Comment (gerardsn) [inline vcr/credential/validator.go:291]: missing test coverage on several error cases

**Review by gerardsn — CHANGES_REQUESTED** [review]
doesn't complete CRL checks mentioned in referenced issue

File: vcr/credential/validator.go [hunk 5]
```diff
@@ -253,3 +256,89 @@ func validateNutsCredentialID(credential vc.VerifiableCredential) error {
...
+		return fmt.Errorf("invalid did:x509 policy")
+	}
+	for _, policy := range policies[1:] {
+		policySplit := strings.Split(policy, ":")
+		if len(policySplit) < 2 {
+			return fmt.Errorf("invalid did:x509 policy '%s'", policy)
+		}
+		policyName := policySplit[0]
+		policyMap[policyName] = make(map[string]string)
+		for i := 1; i < len(policySplit); i += 2 {
```
Comment (reinkrul) [inline vcr/credential/validator.go:311]: could check for `modulo 2 == 1`. If not, it's invalid?

**Review by reinkrul — COMMENTED** [review]
Looks fine aside from Gerard's comments

**Review by gerardsn — APPROVED** [review]
tests fail on #3625

**Pull request merged**